### PR TITLE
Add `beforeRetry` hook

### PR DESCRIPTION
**Pull request by lambdalisue**
Add `beforeRetry` hook just like [got](https://github.com/sindresorhus/got#hooksbeforeretry). I needed it to make sure that an access token is fresh prior to retry.

#### Example usage

```typescript
export const client = ky.extend({
  hooks: {
    beforeRequest: [
      async function(input, options) {
        if (store.refreshRequired) {
          await store.refresh();
        }
        if (store.isAuthorized) {
          (options.headers as Headers).set(
            'X-Authorization',
            `Bearer ${store.accessToken}`,
          );
        }
      },
    ],
    beforeRetry: [
      async function(input, options, errors, retryCount) {
        if (store.refreshRequired) {
          await store.refresh();
        }
        if (store.isAuthorized) {
          (options.headers as Headers).set(
            'X-Authorization',
            `Bearer ${store.accessToken}`,
          );
        }
      },
    ],
  },
});
```

### CODE REVIEW

File: test/hooks.js [hunk 1]
```diff
@@ -245,3 +246,49 @@ test('`afterResponse` hook gets called even if using body shortcuts', async t =>
...
+					options.headers.set('unicorn', fixture);
+				}
+			]
+		}
+	}).text(), fixture);
+
+	await server.close();
+});
+
+test('beforeRetry hook allows modifications of non initial requests', async t => {
```
Comment (sindresorhus) [inline test/hooks.js]:
```suggestion
test('beforeRetry hook allows modifications of non-initial requests', async t => {
```

File: test/hooks.js [hunk 2]
```diff
@@ -245,3 +246,49 @@ test('`afterResponse` hook gets called even if using body shortcuts', async t =>
...
+			response.end(request.headers.unicorn);
+		} else {
+			response.sendStatus(408);
+		}
+	});
+
+	t.is(await ky.get(server.url, {
+		hooks: {
+			beforeRetry: [
+				options => {
```
Comment (sindresorhus) [inline test/hooks.js]: Test the other arguments too, and that the retryCount is correct.

File: test/hooks.js [hunk 3]
```diff
@@ -245,3 +246,49 @@ test('`afterResponse` hook gets called even if using body shortcuts', async t =>
...
+					options.headers.set('unicorn', fixture);
+				}
+			]
+		}
+	}).text(), fixture);
+
+	await server.close();
+});
```
Comment (sindresorhus) [inline test/hooks.js]: Make sure you test as much as done for `beforeRetry` in Got: https://github.com/sindresorhus/got/blob/master/test/hooks.ts

**Comment by lambdalisue** [comment]
I'll wait https://github.com/sindresorhus/ky/pull/163

**Comment by sindresorhus** [comment]
@lambdalisue Can you fix the merge conflict and finish this PR?

**Comment by lambdalisue** [comment]
It was a bit difficult to keep previous commits so I rebased from current master and force-pushed.
Please re-review.

**Comment by lambdalisue** [comment]
@sindresorhus So any updates?

**Comment by szmarczak** [comment]
Travis is very slow, so this is why Puppeteer fails. Just ignore it. The PR looks good.

File: index.d.ts [hunk 4]
```diff
@@ -39,6 +46,15 @@ export interface Hooks {
 	*/
 	beforeRequest?: BeforeRequestHook[];
 
+	/**
+	Before the request is retried.
+
+	This hook enables you to modify the request right before retry. The hook function receives the `input`, `options`, `error` and `retryCount` arguments. You could, for example, modify `options.headers` here.
```
Comment (sindresorhus) [inline index.d.ts]: This should have the exact same text and example as the readme. This makes for a better TS experience, and having them in sync makes it easier to maintain both.

// @szmarczak

Comment (sindresorhus) [reply index.d.ts]: Looks like there's more stuff that's out of sync: https://github.com/sindresorhus/ky/commit/8ab3886a4b57ccf8e6c2ac55a8046c82f3ec211d

Comment (szmarczak) [reply index.d.ts]: Sorry, forgot about syncing these. There should be a tool that converts automatically TS docs into markdown, so we don't have to update the same info twice.

Comment (sindresorhus) [reply index.d.ts]: Ideally, an GitHub Action, but for now, we just have to try to make sure all PRs update index.d.ts with both methods and docs and examples.

Comment (sholladay) [reply index.d.ts]: I think a PR template with a checklist would be a good idea until it can be automated.

**Pull request merged**